### PR TITLE
View log

### PR DESCRIPTION
**Pull request by rossdotparker**
**Description**
Displays a filterable table of gibbonLog entries, as a new action in Admin > System Admin

**Motivation and Context**
Roman made me do it ; )

**How Has This Been Tested?**
Locally

**Screenshots**
Screenshot 2020-05-06 at 4 08 14 PM

### CODE REVIEW

File: modules/System Admin/logs_view.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        returnProcess($guid, $_GET['return'], null, $returns);
+    }
+
+    $ip = isset($_GET['ip'])? $_GET['ip'] : '';
+    $title = isset($_GET['title'])? $_GET['title'] : '';
+    $gibbonPersonID = isset($_GET['gibbonPersonID'])? $_GET['gibbonPersonID'] : '';
+
+    echo '<h2>';
+    echo __('Filters');
+    echo '</h2>';
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:48]:
```suggestion
```

File: modules/System Admin/logs_view.php [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    $form = Form::create('filter', $_SESSION[$guid]['absoluteURL'].'/index.php', 'get');
+
+    $form->setFactory(DatabaseFormFactory::create($pdo));
+    $form->setClass('noIntBorder fullWidth');
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:53]:
```suggestion
    $form->setTitle(__('Filters'));
    $form->setClass('noIntBorder fullWidth');
```

File: modules/System Admin/logs_view.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        $row->addTextField('ip')->setValue($ip);
+
+    $row = $form->addRow();
+        $row->addSearchSubmit($gibbon->session, __('Clear Search'));
+
+    echo $form->getOutput();
+
+    echo '<h2>';
+    echo __('Data');
+    echo '</h2>';
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:78]:
```suggestion
```

File: modules/System Admin/logs_view.php [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    $logs = $logGateway->queryLogs($criteria);
+
+    $table = DataTable::createPaginated('logView', $criteria);
+
+    // $table->addMetaData('filterOptions', [
+    //     'type:core'       => __('Type').': '.__('Core'),
+    //     'type:additional' => __('Type').': '.__('Additional'),
+    //     'active:Y' => __('Active').': '.__('Yes'),
+    //     'active:N' => __('Active').': '.__('No'),
+    // ]);
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:98]:
```suggestion
```

File: modules/System Admin/logs_view.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        ->sortBy('timestamp', 'DESC')
+        ->filterBy('ip', $ip)
+        ->filterBy('title', $title)
+        ->filterBy('gibbonPersonID', $gibbonPersonID)
+        ->fromPOST();
+
+    $logs = $logGateway->queryLogs($criteria);
+
+    $table = DataTable::createPaginated('logView', $criteria);
+
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:92]:
```suggestion
    $table->setTitle(__('Data'));
```

File: modules/System Admin/logs_view.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    //     'active:N' => __('Active').': '.__('No'),
+    // ]);
+
+    $table->addExpandableColumn('comment')
+        ->format(function($log) {
+            $array = unserialize($log['serialisedArray']);
+
+            $details = '';
+            if (count($array) > 0) {
+                $details = "<table class='smallIntBorder' style='width:100%;'>";
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:106]: 🤔

File: src/Domain/System/LogGateway.php [hunk 7]
```diff
@@ -37,14 +37,53 @@ class LogGateway extends QueryableGateway
     private static $primaryKey = 'gibbonLogID';
 
     private static $searchableColumns = ['title'];
-    
+
+    /**
+     * Queries the list for the Manage Modules page.
```
Comment (SKuipers) [inline src/Domain/System/LogGateway.php:42]:
```suggestion
     * Queries the list of System logs.
```

File: modules/System Admin/logs_view.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    echo __('Filters');
+    echo '</h2>';
+
+    $form = Form::create('filter', $_SESSION[$guid]['absoluteURL'].'/index.php', 'get');
+
+    $form->setFactory(DatabaseFormFactory::create($pdo));
+    $form->setClass('noIntBorder fullWidth');
+    $form->addHiddenValue('q', '/modules/'.$_SESSION[$guid]['module'].'/logs_view.php');
+
+    $data = array();
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:56]:
```suggestion
```

File: modules/System Admin/logs_view.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+    $sql = "SELECT DISTINCT title AS value, title AS name FROM gibbonLog ORDER BY title";
+    $row = $form->addRow();
+        $row->addLabel('title', __('Title'));
+        $row->addSelect('title')->fromQuery($pdo, $sql, $data)->selected($title)->placeholder();
```
Comment (SKuipers) [inline modules/System Admin/logs_view.php:60]:
```suggestion
        $row->addSelect('title')->fromQuery($pdo, $sql)->selected($title)->placeholder();
```

File: modules/System Admin/logs_view.php (hunk 6 above)
Comment (rossdotparker) [reply modules/System Admin/logs_view.php:95]: You don't like my crazy table-within-a-table? I copied it from Browse Library. Open to better ways to do this!

**Pull request merged**